// skyerp/debug/logger.js
var util = require( 'util' );

require( "skyerp/utils/inherits.js" );

function dispatch( fn, args ) {
    fn = ( typeof fn == "function" ) ? fn : window[ fn ]; // Allow fn to be a function object or the name of a global function
    return fn.apply( this, args || [ ] ); // args is optional, use an empty array by default
}

function log( format, args ) {
    var res = dispatch( util.format, args );
    var s = util.format( format, res );
    console.log( s );
}

/**
 * @class osurvey.Logger
 * @extends Object
 *
 * Logger de l'application
 *
 */
var Logger = Object.subClass( {
    /**
     * Constructeur
     */
    init: function ( ) {

    },

    /**
     * Ajout d'un log de debug
     */
    debug: function ( ) {
        log( "DEBUG %s", arguments );
    },

    /**
     * Ajout d'un log d'information
     */
    info: function ( arg ) {
        log( "INFO %s", arguments );
    },

    /**
     * Ajout d'un log de warning
     */
    warning: function ( arg ) {
        log( "WARNING %s", arguments );
    },

    /**
     * Ajout d'un log d'erreur
     */
    error: function ( arg ) {
        log( "ERROR %s", arguments );
    }
} );

// Export du module
module.exports = new Logger( );
